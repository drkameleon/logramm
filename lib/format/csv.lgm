/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/csv.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

import core;
import core.string;

//================================================
// Module definition
//================================================

Csv :: 
{
	//================================================
	// Functions
	//================================================

	parse(csv) # [
		.params		: "string", 
		.returns    : "any", 
		.help		: "Processes csv string and returns corresponding object"
	] => 
	{
		lines = csv.split("\n");
		lines = lines.map("trim($)").filter("$!=\"\"");

		header = lines[0];
		headers = header.split(",").map("trim($)");

		ret = [];

		@ line->lines : 
		{
			line!=lines.first() : {

				parts = line.split(",").map("trim($)").filter("$!=\"\"");

				(count(parts)==count(headers)) :
				{
					row = [:];

					@ i,part->parts : row += [ headers[i] : part ];

					ret += row;
				}
			}
		}

		==> ret;
	}

	generate(object) # [
		.params		: "array",
		.returns	: "string",
		.help		: "Processes input and returns corresponding Json as string"
	] => 
	{
		ret = "";

		headers = object.first().keys().map("trim($)");
		
		ret += headers.join(",") + "\n";

		@ row->object : ret += row.values().join(",") + "\n";

		==> ret;
	}
	
}

