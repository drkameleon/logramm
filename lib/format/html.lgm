/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/html.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

import core;

import core.string;

//================================================
// Module definition
//================================================

Html :: 
{

	//================================================
	// Functions
	//================================================

	external parse(markup) # [
		.params		: "string",
		.returns    : "string",
		.help		: "Converts HTML/LGMX template to HTML and returns result"
	] => "html.parse";
	
}
