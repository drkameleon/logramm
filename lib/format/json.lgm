/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/json.lgm
 **********************************************************/

//================================================
// Module definition
//================================================

Json :: 
{

	//================================================
	// External D-bindings
	//================================================

	external parse(json) # [
		.params		: "string", 
		.returns    : "any", 
		.help		: "Processes json string and returns corresponding object"
	] => "json.parse";

	external generate(object) # [
		.params		: "any",
		.returns	: "string",
		.help		: "Processes input and returns corresponding Json as string"
	] => "json.generate";

}

