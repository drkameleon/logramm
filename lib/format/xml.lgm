/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/xml.lgm
 **********************************************************/

//================================================
// Imported modules
//================================================

//================================================
// Module definition
//================================================

Xml :: 
{

	//================================================
	// External D-bindings
	//================================================

	external check(xml) # [
		.params		: "string", 
		.returns    : "boolean", 
		.help		: "Checks an XML document, as string, for well-formedness and returns true if successful"
	] => "xml.check";

}

