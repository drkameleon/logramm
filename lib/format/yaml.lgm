/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** lib/format/yaml.lgm
 **********************************************************/

//================================================
// Module definition
//================================================

Yaml :: 
{

	//================================================
	// External D-bindings
	//================================================

	external parse(yaml) # [
		.params		: "string", 
		.returns    : "any", 
		.help		: "Processes YAML string and returns corresponding object"
	] => "yaml.parse";

	external generate(object) # [
		.params		: "any",
		.returns	: "string",
		.help		: "Processes input and returns corresponding YAML as string"
	] => "yaml.generate";

}

