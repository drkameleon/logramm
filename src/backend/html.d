/**********************************************************
 **
 ** LOGRAMM
 ** Interpreter
 ** 
 ** (c) 2009-2014, Dr.Kameleon
 **
 **********************************************************
 ** src/backend/html.d
 **********************************************************/

module backend.html;

//================================================
// Imports
//================================================

import std.array;
import std.conv;
import std.json;
import std.stdio;
import std.string;

import value;

//================================================
// Class
//================================================

class LGM_Html
{
	static Value parse(Value[] v)
	{
		string str = v[0].content.s;

		string buffer = "";
		string ret = "";
		bool codeMode = false;
		string slice;
		string[] lines;
		bool toOut = false; 

		for (ulong i=0; i<str.length; i++)
		{
			buffer ~= str[i];
			slice = str[i..$];

			if ((std.algorithm.startsWith(slice,"<%")) && !codeMode)
			{
				if (std.algorithm.startsWith(slice,"<%=")) toOut = true;
				else toOut = false;

				lines = buffer[0..$-1].splitter("\n").array();
				foreach (string line; lines)
				{
					if (line!="")
						ret ~= "out '" ~ line.replace("'","\\'") ~ "';\n";
				}
				buffer = "";
				codeMode = true;
				i+= 1 + toOut;
			}

			else if ((std.algorithm.startsWith(slice,"%>")==1) && codeMode)
			{
				lines = buffer[0..$-1].splitter("\n").array();
				foreach (string line; lines)
				{
					if (!toOut)
						ret ~= line.strip() ~ "\n";
					else
						ret ~= "out " ~ line.strip() ~ ";\n";
				}

				buffer = "";
				codeMode = false;
				i+=1;
			}

		}
		buffer ~= str[str.length-1];
		lines = buffer[0..$-1].splitter("\n").array();
		foreach (string line; lines)
		{
			if (line!="")
				ret ~= "out '" ~ line ~ "';\n";
		}

		return new Value(ret);
	}

}
