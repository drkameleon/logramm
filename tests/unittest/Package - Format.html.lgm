import format.html;

html = "<html>
	<head>
		<title><% out('PAGE TITLE'); %></title>
	</head>
	<body>
		<h1><% out('HEADER'); %></h1>
		<div id=\"articles\">
			<% 
				articles = [ 
					[.title:'one',.content:'one-content'],
					[.title:'two',.content:'two-content']
				];
			%>
			<% @ article -> articles : { %>
				<div class=\"article\">
					<h2><% out article.title; %></h2>
					<p class=\"content\">
						<% out article.content; %>
					</p>
				</div>
			<% } %>
		</div>
	</body>
</html>";

out "Parsed ::";
out "=================";
out parse(html);

out "Final ::";
out "=================";
exec parse(html);