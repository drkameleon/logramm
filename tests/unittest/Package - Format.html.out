Parsed ::
=================
out '<html>';
out '	<head>';
out '		<title>';
out('PAGE TITLE');
out '</title>';
out '	</head>';
out '	<body>';
out '		<h1>';
out('HEADER');
out '</h1>';
out '		<div id="articles">';
out '			';

articles = [
[.title:'one',.content:'one-content'],
[.title:'two',.content:'two-content']
];

out '			';
@ article -> articles : {
out '				<div class="article">';
out '					<h2>';
out article.title;
out '</h2>';
out '					<p class="content">';
out '						';
out article.content;
out '					</p>';
out '				</div>';
out '			';
}
out '		</div>';
out '	</body>';
out '</html>';

Final ::
=================
<html>
	<head>
		<title>
PAGE TITLE
</title>
	</head>
	<body>
		<h1>
HEADER
</h1>
		<div id="articles">
			
			
				<div class="article">
					<h2>
one
</h2>
					<p class="content">
						
one-content
					</p>
				</div>
			
				<div class="article">
					<h2>
two
</h2>
					<p class="content">
						
two-content
					</p>
				</div>
			
		</div>
	</body>
</html>